kafka-connect-s3/src/test/java/io/confluent/connect/s3/DataWriterAvroTest.java

```diff
@@ -246,9 +254,152 @@ public void testPreCommit() throws Exception {
   }
 
   @Test
+  public void testPreCommitOnSchemaIncompatibilityRotation() throws Exception {
+    localProps.put(S3SinkConnectorConfig.FLUSH_SIZE_CONFIG, "2");
+    setUp();
+
+    task = new S3SinkTask(connectorConfig, context, storage, partitioner, format, SYSTEM_TIME);
+    List<SinkRecord> sinkRecords = createRecordsWithAlteringSchemas(2, 0);
+
+    // Perform write
+    task.put(sinkRecords);
+
+    Map<TopicPartition, OffsetAndMetadata> offsetsToCommit = task.preCommit(null);
+
+    long[] validOffsets = {1, -1};
+
+    verifyOffsets(offsetsToCommit, validOffsets, context.assignment());
+
+    task.close(context.assignment());
+    task.stop();
+  }
+
+  @Test
+  public void testPreCommitOnRotateTime() throws Exception {
+    // Do not roll on size, only based on time.
+    localProps.put(S3SinkConnectorConfig.FLUSH_SIZE_CONFIG, "1000");
+    localProps.put(
+        S3SinkConnectorConfig.ROTATE_INTERVAL_MS_CONFIG,
+        String.valueOf(TimeUnit.HOURS.toMillis(1))
+    );
+    setUp();
+
+    // Define the partitioner
+    TimeBasedPartitioner<FieldSchema> partitioner = new TimeBasedPartitioner<>();
+    parsedConfig.put(PartitionerConfig.PARTITION_DURATION_MS_CONFIG, TimeUnit.DAYS.toMillis(1));
+    parsedConfig.put(
+        PartitionerConfig.SCHEMA_GENERATOR_CLASS_CONFIG,
+        TimeBasedSchemaGenerator.class
+    );
+    parsedConfig.put(
+        PartitionerConfig.TIMESTAMP_EXTRACTOR_CLASS_CONFIG,
+        TopicPartitionWriterTest.MockedWallclockTimestampExtractor.class.getName()
+    );
+    partitioner.configure(parsedConfig);
+
+    MockTime time = ((TopicPartitionWriterTest.MockedWallclockTimestampExtractor) partitioner
+        .getTimestampExtractor()).time;
+    // Bring the clock to present.
+    time.sleep(SYSTEM.milliseconds());
+
+    List<SinkRecord> sinkRecords = createRecordsWithTimestamp(
+        3,
+        0,
+        Collections.singleton(new TopicPartition(TOPIC, PARTITION)),
+        time
+    );
+
+    task = new S3SinkTask(connectorConfig, context, storage, partitioner, format, time);
+
+    // Perform write
+    task.put(sinkRecords);
+
+    Map<TopicPartition, OffsetAndMetadata> offsetsToCommit = task.preCommit(null);
+
+    long[] validOffsets1 = {-1, -1};
+
+    verifyOffsets(offsetsToCommit, validOffsets1, context.assignment());
+
+    // 2 hours
+    time.sleep(TimeUnit.HOURS.toMillis(2));
+
+    long[] validOffsets2 = {3, -1};
+
+    // Rotation is only based on rotate.interval.ms, so I need at least one record to trigger flush.
+    task.put(sinkRecords.subList(0, 1));
```

Comment:
This looks a bit weird in the test since you're reusing a record with an old offset. This happens to work because there's no checks on the offsets, but it's actually exercising a scenario that should never happen. You could also just allocate more than you need and use `subList` on each call to `task.put`